File: src/Software Engineer Prepare Kit/02/solution.go
package main

func FindSmallestMissingPositive(orderNumbers []int32) int32 {
	length := int32(len(orderNumbers))

	for i := int32(0); i < length; {
		number := orderNumbers[i]

		if number > 0 && number <= length && orderNumbers[number-1] != number {
			orderNumbers[i], orderNumbers[number-1] = orderNumbers[number-1], orderNumbers[i]
		} else {
			i++
		}
	}

	for i := range length {
		if orderNumbers[i] != i+1 {
			return i + 1
		}
	}

	return length + 1
}

File: src/Software Engineer Prepare Kit/02/solution.ts
export default function findSmallestMissingPositive(orderNumbers: number[]): number {
	let i = 0
	while (i < orderNumbers.length) {
		const number = orderNumbers[i]!

		if (number > 0 && number <= orderNumbers.length && orderNumbers[number - 1] !== number) {
			;[orderNumbers[i], orderNumbers[number - 1]] = [orderNumbers[number - 1]!, orderNumbers[i]!]
		} else {
			i++
		}
	}

	for (let i = 0; orderNumbers.length; i++) {
		if (orderNumbers[i] != i + 1) return i + 1
	}

	return orderNumbers.length + 1
}
